smtp/email_service.go

```diff
@@ -0,0 +1,106 @@
+package smtp
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"log"
+	"log/slog"
+	"strings"
+
+	"github.com/a-h/templ"
+	"github.com/prior-it/apollo/config"
+	"github.com/prior-it/apollo/core"
+	"gopkg.in/gomail.v2"
+)
+
+type EmailService struct {
+	from          string
+	d             *gomail.Dialer
+	notifications *core.EmailAddress
+}
+
+func NewEmailService(cfg config.EmailConfig) (*EmailService, error) {
+	d := gomail.NewDialer(
+		cfg.Host,
+		cfg.Port,
+		cfg.User,
+		cfg.Password,
+	)
+
+	s := EmailService{
+		from: cfg.From,
+		d:    d,
+	}
+
+	address, err := core.NewEmailAddress(cfg.Notifications)
+	switch {
+	case errors.Is(err, core.ErrEmailAddressEmpty):
+		slog.Info("Notification e-mail address empty, notifications will not be sent")
+	case err != nil:
+		return nil, err
+	default:
+		s.notifications = &address
+	}
+
+	return &s, nil
+}
+
+// SendEmail will build and send a basic e-mail message containing both a HTML template version (optional) as well as a plaintext alternative (required).
+// This will open a new server connection and immediately close it after sending the e-mail.
+func (s *EmailService) SendEmail(
+	ctx context.Context,
+	address core.EmailAddress,
+	subject string,
+	template *templ.Component,
+	plaintextMessage string,
+) error {
+	m := gomail.NewMessage()
+
+	m.SetHeader("From", s.from)
+	m.SetHeader("To", address.String())
+	m.SetHeader("Subject", subject)
+	m.SetBody("text/plain", plaintextMessage)
+
+	if template != nil {
+		var builder strings.Builder
+		err := (*template).Render(ctx, &builder)
+		if err != nil {
+			log.Fatal(err)
+		}
+		m.AddAlternative("text/html", builder.String())
+	}
+
+	return s.d.DialAndSend(m)
+}
+
+// SendNotification will send a specific plain-text notification to the configured notification
+// address.
+func (s *EmailService) SendNotification(
```

Comment:
Begrijp ik goed dat dit dan ook ondersteuning geeft voor pushnotificaties naar bv browsers? Valt dat onder 'email service'?

Reply:
Anderzijds, dat lijkt ook mail. Ik neem aan dat het eerder gaat over iets als 'serviceberichten' ipv volledige mails

Reply:
Inderdaad, het is een MailService dus die notifications zijn mails en geen push notifications 😅 in Wachtplanner zit er zo een voorbeeld van een notificatie als iemand zich nieuw geregistreerd heeft op de nieuwsbrief. Daarom ook dat die geen templates aanvaarden, dit is bedoeld om heel eenvoudige notificaties naar onszelf te sturen per mail (je kan het to address ook niet aanpassen bv, dat zit enkel in config)